NetKet/Graph/hypercube.hpp

```diff
@@ -46,33 +50,268 @@ class Hypercube : public AbstractGraph {
+    return detail::AdjacencyListFromEdges(Edges(), Nsites());
+  }
+
+  bool IsBipartite() const override { return !pbc_ || L_ % 2 == 0; }
+
+  bool IsConnected() const override { return true; }
+
+  // Returns map of the edge and its respective color
+  const ColorMap &EdgeColors() const override { return colors_; }
+
+  std::vector<std::vector<int>> SymmetryTable() const override {
+    return symm_table_;
+  }
+
+ private:
+  // Given the L, ndim, and pbc, returns (nsites, edges).
+  static std::tuple<int, std::vector<Edge>> BuildEdges(int const L,
+                                                       int const ndim,
+                                                       bool pbc) {
+    assert(L >= (1 + static_cast<int>(pbc)) && ndim >= 1 &&
+           "Bug! Must hold by construction");
+    // NOTE: a double has 53 bits matissa, while an int only 32, so the
+    // following should work fine if we assume that nsites_ fits into an int
+    auto const _n_sites = std::pow(L, ndim);
+    assert(_n_sites == std::trunc(_n_sites) && "Bug! Inexact arithmetic");
+    auto const n_sites = static_cast<int>(_n_sites);
+
+    std::vector<Edge> edges;
+    edges.reserve(static_cast<std::size_t>(n_sites));
+    std::vector<int> coord(ndim, 0);
+    auto const max_pos = L - 1;
+    auto site = 0;
+    do {
+      for (auto i = 1, dim = 0; dim < ndim; ++dim, i *= L) {
+        // NOTE(twesterhout): Hoping that any normal optimising compiler will
+        // move the if (pbc_) dispath to outside the while loop...
+        if (pbc) {
+          if (coord[dim] == max_pos) {
+            edges.push_back({site - i * (L - 1), site});
+          } else {
+            edges.push_back({site, site + i});
+          }
+        } else {
+          if (coord[dim] != max_pos) {
+            edges.push_back({site, site + i});
+          }
+        }
+      }
+      ++site;
+    } while (next_variation(coord.rbegin(), coord.rend(), max_pos));
+    assert(site == n_sites && "Bug! Postcondition violated");
+    return std::tuple<int, std::vector<Edge>>{n_sites, std::move(edges)};
+  }
+
+  static int Coord2Site(std::vector<int> const &coord, int const L) noexcept {
+    assert(L >= 0);
+    auto site = 0;
+    auto scale = 1;
+    for (auto const i : coord) {
+      site += scale * i;
+      scale *= L;
+    }
+    return site;
+  }
+
+  static std::vector<std::vector<int>> BuildSymmTable(int const L,
+                                                      int const n_dim, bool pbc,
+                                                      int const n_sites) {
+    if (!pbc) return {};
+    assert(L >= 2 && n_dim >= 1 &&
+           "Bug! This should be the case by construction.");
+    std::vector<std::vector<int>> symm_table;
+    symm_table.reserve(static_cast<std::size_t>(n_dim * (L - 1)));
+    std::vector<int> coord(static_cast<std::size_t>(n_dim));
+    for (auto dim = 0; dim < n_dim; ++dim) {
+      for (auto rotate_by = 1; rotate_by < L; ++rotate_by) {
```

Comment:
@twesterhout , here you should include the identity, otherwise this gives wrong results for symmetric machines.
I.e. ```rotate_by=0;...``` 
I am correcting this now

Reply:
In fact overall this generates wrong/less symmetries, especially in higher dimension. I am reverting to the old version